pkg/controller/bastion/actuator.go

```diff
@@ -0,0 +1,246 @@
+	alicloudApi "github.com/gardener/gardener-extension-provider-alicloud/pkg/apis/alicloud"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/utils"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/vpc"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/bastion"
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// SSHPort is the default SSH Port used for bastion ingress firewall rule
+	SSHPort = "22/22"
+)
+
+type actuator struct {
+	client            client.Client
+	logger            logr.Logger
+	alicloudECSClient alicloudclient.ECS
+	alicloudVPCClient alicloudclient.VPC
+}
+
+func newActuator() bastion.Actuator {
+	return &actuator{
+		logger: logger,
+	}
+}
+
+func (a *actuator) InjectClient(client client.Client) error {
+	a.client = client
+	return nil
+}
+
+func getAliCloudECSClient(ctx context.Context, client client.Client, opt *Options) (alicloudclient.ECS, error) {
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, client, &opt.SecretReference)
+	if err != nil {
+		return nil, err
+	}
+	return alicloudclient.NewClientFactory().NewECSClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+}
+
+func getAliCloudVPCClient(ctx context.Context, client client.Client, opt *Options) (alicloudclient.VPC, error) {
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, client, &opt.SecretReference)
+	if err != nil {
+		return nil, err
+	}
+	return alicloudclient.NewClientFactory().NewVPCClient(opt.Region, credentials.AccessKeyID, credentials.AccessKeySecret)
+}
+
+func getInstance(a *actuator, opt *Options) (*ecs.DescribeInstancesResponse, error) {
+	request := ecs.CreateDescribeInstancesRequest()
+	request.InstanceName = opt.BastionInstanceName
+	return a.alicloudECSClient.DescribeInstances(request)
+}
+
+func getSecurityGroup(a *actuator, opt *Options, name string) (*ecs.DescribeSecurityGroupsResponse, error) {
+	request := ecs.CreateDescribeSecurityGroupsRequest()
+	request.RegionId = opt.Region
+	request.SecurityGroupName = name
+	return a.alicloudECSClient.DescribeSecurityGroups(request)
+}
+
+func getClusterProviderConfig(bytes []byte) (*alicloudApi.CloudProfileConfig, error) {
+	cloudProfileConfig := &alicloudApi.CloudProfileConfig{}
+
+	err := json.Unmarshal(bytes, cloudProfileConfig)
+	if err != nil {
+		return nil, errors.New("failed to parse json for cloud ProfileConfig")
+	}
+	return cloudProfileConfig, nil
+}
+
+func getClusterImageID(cloudProfileConfig alicloudApi.CloudProfileConfig, name, version, region string) string {
+	for _, machineImage := range cloudProfileConfig.MachineImages {
+		if machineImage.Name == name {
+			for _, machine := range machineImage.Versions {
+				if machine.Version == version {
+					for _, machineRegionIDMappingImage := range machine.Regions {
```

Comment:
actually, the value return from `machineRegionIDMappingImage.ID` is image ID purpose. to avoid misleading I change it to `machineRegionIDMappingImage`.  let me know if need to rename it to `region` and return `region.ID` as below. 
original information https://github.com/tedteng/gardener-extension-provider-alicloud/blob/0e74b1beaa2097ed5f136b3c755c1ae8d977515e/pkg/apis/alicloud/types_cloudprofile.go#L48-L54
```

Regions []RegionIDMapping 

// RegionIDMapping is a mapping to the correct ID for the machine image in the given region.
type RegionIDMapping struct {
	// Name is the name of the region.
	Name string
	// ID is the id of the image.
	ID string
}
```

Reply:
https://github.com/tedteng/gardener-extension-provider-alicloud/blob/0e74b1beaa2097ed5f136b3c755c1ae8d977515e/pkg/apis/alicloud/types_cloudprofile.go#L45 here it's also just named `Regions`. I'd still vote for `region`

Reply:
committed in https://github.com/gardener/gardener-extension-provider-alicloud/commit/14aef4bc5adeaeeddeb68069f49315f3115c33e4